### PR TITLE
Python Implementation of Ubisoft's Clever-Initiative coding challenge

### PR DESCRIPTION
**Pull request by andykdy**
Hello, I have implemented a solution for your challenge using Python. 

### Part 1: Diff Logger
Instruction : **python -p 1 -f \<location of diff file\>**

There is a regex pattern for each stat being logged. 
Each line of the diff file is checked using each pattern. 
Depending on the stat, a counter is incremented, a value is appended to a list, or both in the case of function calls.

#### Known problems:
functionCalls will catch offsets from assembly code as a function call. 
e.g. In the line  
`-    "movl        8(%ebp),%eax       \n\t"`
'8' will be noted as a function call which isn't correct. 

#### Possible improvement:
I could see a modular approach to this question where each stat being recorded could be compartmentalized with its own regex pattern and method of storage.  

For example,
Stat Name : Lineadded
Regex Pattern : `^(\+).*`
When Found: Increment

This way, the user could create new stats easily without having to change the code allowing extension but not modification. 

### Part 2: Variable declarations in Abstract Syntax Tree
Instruction : **python -p 2 -f \<astChallenge.json\>**

All variables are located within a 'VariableDeclaration' node, thus the code traverses through the AST recursively and stores all 'VariableDeclaration' nodes to a list.   
Each 'VariableDeclaration' node in the list is then converted into a tuple. 
The name of the variable is found under VariableDeclarator while the type of the variable is found under PredefinedType.

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
```
Comment (lbajolet) [inline .idea/vcs.xml]: I'm not sure including the `vcs.xml` from your IDE is required here, we're interested in the code more than the setup

File: diff_result.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from collections import defaultdict
+class DiffResult:
+    def __init__(self):
+        self.files = []
+        self.regions = 0
+        self.lineAdded = 0
+        self.lineDeleted = 0
+        self.functionCalls = defaultdict(int)
+
+    def toText(self):
```
Comment (lbajolet) [inline diff_result.py]: I'm not a hardcore pythonista, but I thought `camelCase` was strongly discouraged in favour of `snake_case`, if you were rewriting the project why not stick to the standard?

File: ast_parser.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            traverseSearch(child, lookfor, resultList)
+
+# Use traverseSearch() to find...
+# Variable Name found under node VariableDeclarator
+# Variable Type found under node PredefinedType
+def findVal(varNode, lookFor):
+    found = []
+    traverseSearch(varNode, lookFor, found)
+    if found:
+        return found[0]['Children'][0]['ValueText']
```
Comment (lbajolet) [inline ast_parser.py:63]: This could fail with a broken tree, but I guess for this test it's ok

File: main.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import argparse
```
Comment (lbajolet) [inline main.py:1]: Shebang please

File: diffs/diff4.diff [hunk 5]
```diff
@@ -0,0 +1,108 @@
+*TEST FILE FOR DIFF*
```
Comment (lbajolet) [inline diffs/diff4.diff:1]: What's special with this diff file to justify its inclusion? Does it add an edge case or something?

File: diff_result.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        with open('diffResult.txt', 'w') as output:
+            output.write("Files: \n")
```
Comment (lbajolet) [inline diff_result.py:13]: When testing, files are mislabeled as you left the `a/` and `b/` prefixes, most of them are doubled

File: diff_result.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+class DiffResult:
+    def __init__(self):
+        self.files = []
+        self.regions = 0
+        self.lineAdded = 0
+        self.lineDeleted = 0
+        self.functionCalls = defaultdict(int)
+
+    def toText(self):
+        with open('diffResult.txt', 'w') as output:
```
Comment (lbajolet) [inline diff_result.py:12]: This could have been an option to the commandline, but it's fine this way too

**Review by lbajolet — COMMENTED** [review]
Overall it's concise, and the results are close to what is expected (the file list needs a rework)
The runtime looks ok to me, ~400ms for diff1; I guess a full run on all files takes ~450ms, a note on performance might have been nice though.
The commit's contents and messages look off in a few places (160e8e7 especially comes to mind).

Sorry for the delay for the review.

File: main.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        astParser = ASTParser()
+        with open("ast/" + filename) as file:
+            result = astParser.parse(file)
+        result.toText()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-p', '--part', required=True)
+    parser.add_argument('-f', '--filename')
```
Comment (chadykamar) [inline main.py]: `filename` should be `required=True`, same as `part` because running the script without it throws an error.

File: ast_parser.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    def parse(self, file):
+        y = json.load(file)
+        ast_res = ASTResult()
+        root = y['Root']
+        varNodes = []
+        traverseSearch(root, 'VariableDeclaration', varNodes)  # Returns a list of nodes from the AST that are variables
+        ast_res.variableDeclarations = nodeToVar(varNodes)  # Parses each variable node to a variable tuple and returns a list of tuples
+        return ast_res
+
+# Converts each variable node to a variable tuple
```
Comment (chadykamar) [inline ast_parser.py]: In Python, documentation is written like so:

```python
def node_to_var(var_nodes):
"""Converts each variable node to a variable tuple"""
```

File: ast_parser.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+from ast_result import ASTResult
+
+# *Goal*
+# Parse an AST to return all the declared variables in the following format
+# {int}{myInt}
+# {string}{myInt}
+# {Foo}{myFooObject}
+
+
+class ASTParser:
```
Comment (chadykamar) [inline ast_parser.py:11]: `ASTParser` doesn't have any state and only has one method. A function `parse_ast` would work just as well here.

File: ast_parser.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self):
+        print("ASTParser created")
+
+    # Load JSON file into dictionary and parse
+    def parse(self, file):
+        y = json.load(file)
```
Comment (chadykamar) [inline ast_parser.py]: `y` is not a good variable name in this case

File: ast_parser.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        varNodes = []
+        traverseSearch(root, 'VariableDeclaration', varNodes)  # Returns a list of nodes from the AST that are variables
+        ast_res.variableDeclarations = nodeToVar(varNodes)  # Parses each variable node to a variable tuple and returns a list of tuples
+        return ast_res
+
+# Converts each variable node to a variable tuple
+def nodeToVar(varNodes):
+    newList = []
+    for node in varNodes:
+        isArray = []
```
Comment (chadykamar) [inline ast_parser.py]: Variables named `is_something`, `has_something`, etc. are usually used for boolean variables or functions.

File: ast_parser.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        traverseSearch(node, 'ArrayCreationExpression', isArray)
+        varType = findVal(node, 'PredefinedType')
+        varName = findVal(node, 'VariableDeclarator')
+        if isArray:
+            varType += "[]"
+        newList.append((varType, varName))
+    return newList
+
+# Recursive traversal of AST to find a node with Type == lookfor
+# Appends all nodes that match to resultList which is maintained because python is pass by reference
```
Comment (chadykamar) [inline ast_parser.py]: Python is not exactly pass by reference, it's a little more complicated than that.

File: .idea/vcs.xml (hunk 1 above)
Comment (andykdy) [reply .idea/vcs.xml]: I must have mistakenly added these when PyCharm made a suggestion, JetBrain configuration files will be removed.

File: diff_result.py (hunk 2 above)
Comment (andykdy) [reply diff_result.py]: You're right, this is a habit that I should be easy to kick out with the help of linters

File: diffs/diff4.diff (hunk 5 above)
Comment (andykdy) [reply diffs/diff4.diff:1]: In the absence of unit testing I instead opted to have a smaller diff file where I could more easily visualize the number of added lines, deleted lines, and functionCalls. 
It was also helpful when it came to testing certain edge cases as well.
If this was code to be deployed, it would need a suite of unit tests to ensure functionality over time.

File: diff_result.py (hunk 6 above)
Comment (andykdy) [reply diff_result.py:13]: I noticed this but was not sure if this was the intended result or if the common sub-path was to be presented. This will be fixed.

File: diff_result.py (hunk 7 above)
Comment (andykdy) [reply diff_result.py:12]: A valid suggestion, thank you

File: main.py (hunk 8 above)
Comment (andykdy) [reply main.py]: Yes, I agree

File: ast_parser.py (hunk 9 above)
Comment (andykdy) [reply ast_parser.py]: Changes have been made to use docstrings instead, thank you

File: ast_parser.py (hunk 11 above)
Comment (andykdy) [reply ast_parser.py]: I let this slip through my final review process, it has been renamed accordingly

File: ast_parser.py (hunk 12 above)
Comment (andykdy) [reply ast_parser.py]: Yes, I have given the variable a more suitable name

File: ast_parser.py (hunk 10 above)
Comment (andykdy) [reply ast_parser.py:11]: Initially I created a class for each parser to allow them to be extendable. However, I agree that they could easily be a function in main.py

File: ast_parser.py (hunk 13 above)
Comment (andykdy) [reply ast_parser.py]: Is it because Python is "Pass-by-object-reference"? 
I have written "Pass-by-reference" because "Pass-by-reference" and "Pass-by-value" are the more common terms used.

Comment (chadykamar) [reply ast_parser.py]: Correct, whereas most languages pass objects by reference, in Python objects are passed by value of reference. It usually does what you expect, but can sometimes bite you...

File: ast_parser.py (hunk 10 above)
Comment (chadykamar) [reply ast_parser.py:11]: It's ok to keep it in ast_parser.py, it's nicely organized that way.
